faster generation of alternate map matches without limitation

kevinkreiser opened the pull request:
fixes #978 

@ptpt and I essentially implemented exactly what was discussed in the issue. we basically deleted candidates from the states that had paths that didnt differ from already found results. in doing so we greatly reduce the number of times one has to do the viterbi search algorithm. in fact we've removed the whole notion of checking whether a whole path is unique against a previously found result because change should guarantee that a subsequent viterbi search will find the next unique result.

dgearhart commented on Makefile.am:
```diff
@@ -885,8 +885,8 @@ endif
 if DATA_TOOLS
 #some of the tests want tiles of a certain location so lets build those
 test/data/utrecht_tiles: valhalla_build_tiles
-	@echo "Building Utrecht Tiles..." && rm -rf test/data/*utrecht_tiles && ./valhalla_build_tiles --inline-config '{"mjolnir":{"tile_dir":"test/data/tmp_utrecht_tiles","concurrency":1,"logging":{"type":""}}}' test/data/utrecht_netherlands.osm.pbf && mv test/data/tmp_utrecht_tiles test/data/utrecht_tiles
```
was this a mistake? comment out

kevinkreiser replied:
yeah there is other printing stuff in there too. basically the pr isnt ready for review yet sorry. just starting it so that @ptpt and i can collaborate on it

kevinkreiser commented:
ok did some work on this yesterday and today and fixed a couple things thanks to @ptpt 's advice. the only problem remaining that i can see is with being able to tell which paths are unique and not. there is currently an issue when one of the losing candidates paths is from one node to the same node (different edge though) this ends up being a comparison of no edges and so it looks unique. im trying to think about how to make this check logically consistent..

kevinkreiser commented:
ok so ive been chasing my tail on this one for a while now and it turns out to be a pretty complicated series of events.

the current status is, everything is much faster wrt topk computations. but the tests are still failing. the reason the test fails is because the first one that @dgearhart added is getting its 2nd most likely alternate as having a uturn and then the 3rd most likely alternate is the path @dgearhart was expecting.

for the longest time i couldnt figure out why. so this morning about 1.5 hours ago i decided i'd hack the code to just return as many unique results as possible without stopping. heres what duanes input looks like:
![a](https://user-images.githubusercontent.com/697548/32440238-52ce3f06-c2c0-11e7-8ea8-b7cbec665726.png)

the 1st result with the updated code is as expected:
![b](https://user-images.githubusercontent.com/697548/32440248-5b6f9cae-c2c0-11e7-85a1-f3f7f1761d75.png)

the 2nd result is unfortunate:
![c](https://user-images.githubusercontent.com/697548/32440263-6d0dbd9c-c2c0-11e7-92bf-815b0855f733.png)

and the 3rd result is what we were hoping for:
![d](https://user-images.githubusercontent.com/697548/32440275-76eabcc0-c2c0-11e7-930b-37858a5f4eb4.png)

For the longest time i couldnt figure out what of all the myriad changes made it so that intuitive second result was happening 3rd instead of 2nd. So i went back to duane branch and moved my changes 1 by 1 to figure out what was doing it leaving the actual removal of candidates until the last thing. it turned out that change that did it was the change to transition_cost_model.cc where we removed the optimization to only add destination candidates who werent routed to before (ie those who didnt have valid predecessor labels, if im understanding the code correctly).

I think what this means is that the test case is actually just not that good! essentially that 2nd result of a uturn is slightly more likely (given the interplay between transition and emission cost) than the 3rd result. just that we didnt see it before because it wasnt even being considered at all (that optimization that i commented out allowed it to be re-computed for `k = 1`).

I'll chat with @ptpt directly to see what he thinks about my hypothesis here.

kevinkreiser commented:
@ptpt  this is finally ready for review. note that i didnt end up actually making use of removing paths directly (because it was more costly than it was worth). but we could add that if we could find a way to make it faster. in the end we:

1. removed candidates which were entirely redundant
2. changed around some of the topk datastructures so that a few of the common operations were faster
3. removed an optimisation in the viterbi search so that we wouldnt skip computing certain paths. this is needed to allow for removing candidates that are redundant. the test that does 100 random map matches was taking just under 5 seconds on my machine. i'll check what master is and post back. if its slower we can make a setting in viterbi that enables only for requests where `k > 1`

dgearhart commented on src/meili/map_matcher.cc:
```diff
@@ -325,93 +347,242 @@ void MapMatcher::Clear()
   }
+}
+
+// Mark any sub paths (pairs of candidates between two states) that are redundant with this one
```
was this left in in purpose?

dgearhart commented on src/meili/transition_cost_model.cc:
```diff
@@ -135,10 +135,10 @@ TransitionCostModel::UpdateRoute(const StateId& lhs, const StateId& rhs) const
   unreached_stateids.reserve(right_column.size());
   for (const auto& state : right_column) {
-    if (!vs_.Predecessor(state.stateid()).IsValid()) {
+    //if (!vs_.Predecessor(state.stateid()).IsValid()) {
```
should you remove the if and end bracket?

dgearhart commented:
high level review looks good to me - I yield to @ptpt for final review

ptpt commented on valhalla/meili/viterbi_search.h:
```diff
@@ -263,14 +262,13 @@ class NaiveViterbiSearch: public IViterbiSearch
-    // clear current search status to remove possible uses of the removed state ID
-    ClearSearch();
-    return true;
+    return size != column.size();
```
Why is this needed? Isn't it always true?

kevinkreiser replied:
i had this in here because i was debugging and trying to make sure that what was in the `added_states_` was also in the `states_` i think we've proven that that is ok so we can remove this.

ptpt commented on src/meili/topk_search.cc:
```diff
@@ -83,34 +114,32 @@ void EnlargedViterbiSearch::ClonePath(const StateId::Time& time)
+}
 
-  return last_valid_origin;
+StateId TopKSearch::GetOrigin(const StateId& stateid, const StateId& not_found) const
```
👍 for this change!

ptpt commented on src/meili/topk_search.cc:
```diff
@@ -3,14 +3,26 @@
 
+float EnlargedEmissionCostModel::operator()(const StateId& stateid) {
+  // If this was a removed state bail
+  const auto& original_stateid = evs_.GetOrigin(stateid);
```
what if `original_stateid` is invalid? i.e. stateid is origin.

kevinkreiser replied:
if its invalid `IsRemoved` will return false, we will then search for the unordinal id in the cache, if its not found, the invalid original id gets sent to `calculate_cost` which will see its invalid and not use it

ptpt replied:
It won't return the cost for `stateid` if `stateid` is an origin, will it? if it's an origin, then `original_stateid` will be invalid, and then we get -1 from `calculate_cost`.

kevinkreiser replied:
if `stateid` is an origin, then if its not marked as redundant it will look it up in the cache. if its not in the cache it will call `calculate_cost`, see the origin is invalid and then use the model to get the cost

ptpt replied:
I get it. You just check if the original stateid is removed or not.

kevinkreiser replied:
yep!

ptpt commented on valhalla/meili/topk_search.h:
```diff
@@ -32,24 +36,30 @@ class EnlargedTransitionCostModel
-        clone_end_time_(kInvalidTime)
+        clone_end_time_(kInvalidTime),
+        initial_origins_(initial_origins),
+        removed_origins_(removed_origins)
```
`removed_stateids` might be better because it allows removing clones too.

kevinkreiser replied:
only original ids are ever removed because those are the ones we use in `RemoveRendundancies` to check for redundant paths. these are candidates that are found within the `StateContainer`.

we also use this trick to remove clones by checking above for the clones original id in the enlarged emission cost model. this way we dont have to store clones at all, we simply look what the original id was and if the redundancy check marked it then the emission cost is `-1` and the clones path will be removed
